### PR TITLE
Convert AcademicDatesBlock test from Kernel to Unit

### LINKED ISSUE #9877: Convert AcademicDatesBlock test from Kernel to Unit
# Story

As a developer, I want `AcademicDatesBlock`'s test to run as a unit test rather than a kernel test, so that it gives faster feedback and does not depend on a Drupal bootstrap it never uses.

The block's logic is constructor-injected, so it can be covered with mocks alone. The current test installs a module and boots the kernel for no benefit.

# Proposed solution

Move it to a unit test that mocks the injected services. Confirm coverage is at least equivalent to the current kernel test.

# Definition of done

- Test needs no database or module install.
- Coverage equivalent or better.

Related: #8181, #9631

---

🤖 *Drafted by [Claude Code](https://claude.com/claude-code) on behalf of @pyrello.*

### PR DESCRIPTION
**Pull request by pyrello**
## Summary
- Move `AcademicDatesBlock`'s test from `tests/src/Kernel/` to `tests/src/Unit/`. Runtime drops from seconds to ~25ms (no DB, no module install, no bootstrap).
- Inject a translation stub for the `$this->t()` calls in the exercised paths.
- Tighten assertions:
  - Exact placeholder output via `[input, expected]` data-provider pairs.
  - `assertNull` / `assertSame` replace loose `assertEquals` (the old form passed even when values were nulled or coerced).
  - Instance-scoped `getErrors()` replaces the global `FormState::hasAnyErrors()` static.

## Watch-out
`FormState::hasAnyErrors()` reads a process-global static. The kernel test harness silently resets it between tests; `UnitTestCase` does not. Any blockSubmit-style test asserting on `hasAnyErrors()` would leak across tests. Fixed here by switching to `getErrors()` (instance state). Candidate for the #8181 testing conventions.

## Test plan
- [x] `ddev phpunit --filter AcademicDatesBlockTest docroot/modules/custom/uiowa_maui/tests/src/Unit/AcademicDatesBlockTest.php` runs green locally
- [x] Travis green

Closes #9877

---

🤖 *Drafted by [Claude Code](https://claude.com/claude-code) on behalf of @pyrello.*

### CODE REVIEW

**Comment by pyrello** [comment]
Here is a diff of the two files for more easy comparison:
<details>
<summary>Diff</summary>

```diff
diff --git a/docroot/modules/custom/uiowa_maui/tests/src/Kernel/AcademicDatesBlockTest.php b/docroot/modules/custom/uiowa_maui/tests/src/Unit/AcademicDatesBlockTest.php                        
index be1912f5d..209bdc9b6 100644                                                                                                                                                               
--- a/docroot/modules/custom/uiowa_maui/tests/src/Kernel/AcademicDatesBlockTest.php                                                                                                             
+++ b/docroot/modules/custom/uiowa_maui/tests/src/Unit/AcademicDatesBlockTest.php                                                                                                               
@@ -1,22 +1,20 @@                                                                                                                                                                               
 <?php                                                                                                                                                                                          
                                                                                                                                                                                                
-namespace Drupal\Tests\uiowa_maui\Kernel;                                                                                                                                                      
+namespace Drupal\Tests\uiowa_maui\Unit;                                                                                                                                                        
                                                                                                                                                                                                
+use Drupal\Core\Form\FormBuilderInterface;                                                                                                                                                     
 use Drupal\Core\Form\FormState;                                                                                                                                                                
-use Drupal\KernelTests\KernelTestBase;                                                                                                                                                         
+use Drupal\Tests\UnitTestCase;                                                                                                                                                                 
+use Drupal\uiowa_maui\MauiApi;                                                                                                                                                                 
 use Drupal\uiowa_maui\Plugin\Block\AcademicDatesBlock;                                                                                                                                         
                                                                                                                                                                                                
 /**                                                                                                                                                                                            
- * Test description.                                                                                                                                                                           
+ * Tests the AcademicDatesBlock build output and configuration handling.                                                                                                                       
  *                                                                                                                                                                                             
  * @group uiowa_maui                                                                                                                                                                           
+ * @coversDefaultClass \Drupal\uiowa_maui\Plugin\Block\AcademicDatesBlock                                                                                                                      
  */                                                                                                                                                                                            
-class AcademicDatesBlockTest extends KernelTestBase {                                                                                                                                          
-                                                                                                                                                                                               
-  /**                                                                                                                                                                                          
-   * {@inheritdoc}                                                                                                                                                                             
-   */                                                                                                                                                                                          
-  protected static $modules = ['uiowa_maui'];                                                                                                                                                  
+class AcademicDatesBlockTest extends UnitTestCase {                                                                                                                                            
                                                                                                                                                                                                
   /**                                                                                                                                                                                          
    * Mock MAUI service.                                                                                                                                                                        
@@ -28,12 +26,12 @@ class AcademicDatesBlockTest extends KernelTestBase {                                                                                                                       
   /**                                                                                                                                                                                          
    * Mock FormBuilder service.                                                                                                                                                                 
    *                                                                                                                                                                                           
-   * @var \Drupal\Core\Form\FormBuilder|\PHPUnit\Framework\MockObject\MockObject                                                                                                               
+   * @var \Drupal\Core\Form\FormBuilderInterface|\PHPUnit\Framework\MockObject\MockObject                                                                                                      
    */                                                                                                                                                                                          
   protected $formBuilder;                                                                                                                                                                      
                                                                                                                                                                                                
   /**                                                                                                                                                                                          
-   * Fake block plugin configuration.
+   * Fake block plugin definition.                                                                                                                                                             
    *                                                                                                                                                                                           
    * @var string[]                                                                                                                                                                             
    */                                                                                                                                                                                          
@@ -72,64 +70,56 @@ protected function setUp(): void {                                                                                                                                          
       'more_text' => 'View more',                                                                                                                                                              
     ];                                                                                                                                                                                         
                                                                                                                                                                                                
-    $this->maui = $this->getMockBuilder('\Drupal\uiowa_maui\MauiApi')                                                                                                                          
-      ->disableOriginalConstructor()                                                                                                                                                           
-      ->getMock();                                                                                                                                                                             
-                                                                                                                                                                                               
-    $this->maui->expects($this->any())                                                                                                                                                         
-      ->method('getCurrentSession')                                                                                                                                                            
-      ->will($this->returnValue(                                                                                                                                                               
-        (object) [                                                                                                                                                                             
-          'id' => 1,                                                                                                                                                                           
-          'shortDescription' => 'Winter 2020',                                                                                                                                                 
-        ]                                                                                                                                                                                      
-      ));                                                                                                                                                                                      
-                                                                                                                                                                                               
-    $this->maui->expects($this->any())                                                                                                                                                         
-      ->method('getSessionsBounded')                                                                                                                                                           
-      ->will($this->returnValue([                                                                                                                                                              
-        (object) [                                                                                                                                                                             
-          'id' => 1,                                                                                                                                                                           
-          'shortDescription' => 'Winter 2020',                                                                                                                                                 
-        ],                                                                                                                                                                                     
-        (object) [                                                                                                                                                                             
-          'id' => 2,                                                                                                                                                                           
-          'shortDescription' => 'Spring 2021',                                                                                                                                                 
-        ],                                                                                                                                                                                     
-        (object) [                                                                                                                                                                             
-          'id' => 3,                                                                                                                                                                           
-          'shortDescription' => 'Summer 2021',                                                                                                                                                 
-        ],                                                                                                                                                                                     
-      ]));                                                                                                                                                                                     
+    $this->maui = $this->createMock(MauiApi::class);                                                                                                                                           
                                                                                                                                                                                                
-    $this->formBuilder = $this->getMockBuilder('\Drupal\Core\Form\FormBuilder')                                                                                                                
-      ->disableOriginalConstructor()                                                                                                                                                           
-      ->getMock();                                                                                                                                                                             
+    $this->formBuilder = $this->createMock(FormBuilderInterface::class);                                                                                                                       
+    $this->formBuilder->method('getForm')->willReturn([]);                                                                                                                                     
+  }
                                                                                                                                                                                                
-    $this->formBuilder->expects($this->any())                                                                                                                                                  
-      ->method('getForm')                                                                                                                                                                      
-      ->will($this->returnValue([]));                                                                                                                                                          
+  /**                                                                                                                                                                                          
+   * Builds the block under test with a translation stub attached.                                                                                                                             
+   *                                                                                                                                                                                           
+   * @param array $config                                                                                                                                                                      
+   *   The block configuration. Every other constructor argument is identical                                                                                                                  
+   *   across tests, so only the varying config is passed in.                                                                                                                                  
+   *                                                                                                                                                                                           
+   * @return \Drupal\uiowa_maui\Plugin\Block\AcademicDatesBlock                                                                                                                                
+   *   The configured block.                                                                                                                                                                   
+   */                                                                                                                                                                                          
+  protected function makeBlock(array $config): AcademicDatesBlock {                                                                                                                            
+    $block = new AcademicDatesBlock($config, 'uiowa_maui_academic_dates', $this->plugin, $this->maui, $this->formBuilder);                                                                     
+    $block->setStringTranslation($this->getStringTranslationStub());                                                                                                                           
+    return $block;                                                                                                                                                                             
   }                                                                                                                                                                                            
                                                                                                                                                                                                
   /**                                                                                                                                                                                          
-   * Test headline placeholder is replaced properly.                                                                                                                                           
+   * Tests that the @session placeholder is replaced and nothing else is.                                                                                                                      
    *                                                                                                                                                                                           
+   * @covers ::build                                                                                                                                                                           
    * @dataProvider placeholderProvider                                                                                                                                                         
    */                                                                                                                                                                                          
-  public function testHeadlinePlaceholderIsReplaced($placeholder) {                                                                                                                            
+  public function testHeadlinePlaceholderIsReplaced(string $placeholder, string $expected): void {                                                                                             
+    $this->maui->method('getSessionsBounded')->willReturn([                                                                                                                                    
+      (object) ['id' => 1, 'shortDescription' => 'Winter 2020'],                                                                                                                               
+      (object) ['id' => 2, 'shortDescription' => 'Spring 2021'],                                                                                                                               
+      (object) ['id' => 3, 'shortDescription' => 'Summer 2021'],                                                                                                                               
+    ]);                                                                                                                                                                                        
+                                                                                                                                                                                               
     $config = $this->blockConfig;                                                                                                                                                              
     $config['headline'] = $placeholder;                                                                                                                                                        
-    $sut = new AcademicDatesBlock($config, 'uiowa_maui_academic_dates', $this->plugin, $this->maui, $this->formBuilder);                                                                       
+    $sut = $this->makeBlock($config);                                                                                                                                                          
                                                                                                                                                                                                
     $build = $sut->build();                                                                                                                                                                    
-    $this->assertStringContainsString('Winter 2020', $build['heading']['#headline']);                                                                                                          
+    $this->assertSame($expected, $build['heading']['#headline']);                                                                                                                              
   }                                                                                                                                                                                            
    /**                                                                                                                                                                                          
-   * Test headline placeholder.                                                                                                                                                                
+   * Tests validation rejects the @session placeholder with an exposed session.                                                                                                                
+   *                                                                                                                                                                                           
+   * @covers ::blockValidate                                                                                                                                                                   
    */                                                                                                                                                                                          
-  public function testHeadlinePlaceholderCannotBeUsedWithExposedSession() {                                                                                                                    
-    $sut = new AcademicDatesBlock([], 'uiowa_maui_academic_dates', $this->plugin, $this->maui, $this->formBuilder);                                                                            
+  public function testHeadlinePlaceholderCannotBeUsedWithExposedSession(): void {                                                                                                              
+    $sut = $this->makeBlock([]);                                                                                                                                                               
     $form_state = new FormState();                                                                                                                                                             
                                                                                                                                                                                                
     $form_state->setValues([                                                                                                                                                                   
@@ -143,15 +133,16 @@ public function testHeadlinePlaceholderCannotBeUsedWithExposedSession() {                                                                                                 
     ]);                                                                                                                                                                                        
                                                                                                                                                                                                
     $sut->blockValidate([], $form_state);                                                                                                                                                      
-    $this->assertTrue($form_state->hasAnyErrors());                                                                                                                                            
+    $this->assertArrayHasKey('session', $form_state->getErrors());                                                                                                                             
   }                                                                                                                                                                                            
                                                                                                                                                                                                
   /**                                                                                                                                                                                          
-   * Test empty session and category select options are saved as NULL.                                                                                                                         
+   * Tests empty session and category selections are saved as NULL.                                                                                                                            
+   *                                                                                                                                                                                           
+   * @covers ::blockSubmit                                                                                                                                                                     
    */                                                                                                                                                                                          
-  public function testEmptyValuesSavedAsNull() {                                                                                                                                               
-    $sut = new AcademicDatesBlock([], 'uiowa_maui_academic_dates', $this->plugin, $this->maui, $this->formBuilder);                                                                            
-                                                                                                                                                                                               
+  public function testEmptyValuesSavedAsNull(): void {                                                                                                                                         
+    $sut = $this->makeBlock([]);                                                                                                                                                               
     $form_state = new FormState();                                                                                                                                                             
                                                                                                                                                                                                
     $form_state->setValues([                                                                                                                                                                   
@@ -165,17 +156,19 @@ public function testEmptyValuesSavedAsNull() {                                                                                                                            
     ]);                                                                                                                                                                                        
                                                                                                                                                                                                
     $sut->blockSubmit([], $form_state);                                                                                                                                                        
-    $this->assertFalse($form_state->hasAnyErrors());                                                                                                                                           
+    $this->assertEmpty($form_state->getErrors());                                                                                                                                              
     $config = $sut->getConfiguration();                                                                                                                                                        
-    $this->assertEquals(NULL, $config['session']);                                                                                                                                             
-    $this->assertEquals(NULL, $config['category']);                                                                                                                                            
+    $this->assertNull($config['session']);
+    $this->assertNull($config['category']);                                                                                                                                                    
   }                                                                                                                                                                                            
                                                                                                                                                                                                
   /**                                                                                                                                                                                          
-   * Test selected values for session and category.                                                                                                                                            
+   * Tests selected session and category values are preserved.                                                                                                                                 
+   *                                                                                                                                                                                           
+   * @covers ::blockSubmit                                                                                                                                                                     
    */                                                                                                                                                                                          
-  public function testNonEmptyValuesSavedAsNotNull() {                                                                                                                                         
-    $sut = new AcademicDatesBlock([], 'uiowa_maui_academic_dates', $this->plugin, $this->maui, $this->formBuilder);                                                                            
+  public function testNonEmptyValuesSavedAsNotNull(): void {                                                                                                                                   
+    $sut = $this->makeBlock([]);                                                                                                                                                               
     $form_state = new FormState();                                                                                                                                                             
                                                                                                                                                                                                
     $form_state->setValues([                                                                                                                                                                   
@@ -189,49 +182,58 @@ public function testNonEmptyValuesSavedAsNotNull() {                                                                                                                      
     ]);                                                                                                                                                                                        
                                                                                                                                                                                                
     $sut->blockSubmit([], $form_state);                                                                                                                                                        
-    $this->assertFalse($form_state->hasAnyErrors());                                                                                                                                           
+    $this->assertEmpty($form_state->getErrors());                                                                                                                                              
     $config = $sut->getConfiguration();                                                                                                                                                        
-    $this->assertEquals(0, $config['session']);                                                                                                                                                
-    $this->assertEquals(1, $config['category']);                                                                                                                                               
+    $this->assertSame(0, $config['session']);                                                                                                                                                  
+    $this->assertSame(1, $config['category']);                                                                                                                                                 
   }                                                                                                                                                                                            
                                                                                                                                                                                                
   /**                                                                                                                                                                                          
-   * The more link should render if display more link is checked.                                                                                                                              
+   * Tests the more link renders with expected text and URL when enabled.                                                                                                                      
+   *                                                                                                                                                                                           
+   * @covers ::build                                                                                                                                                                           
    */                                                                                                                                                                                          
-  public function testMoreLinkDoesRenderIfSet() {                                                                                                                                              
+  public function testMoreLinkDoesRenderIfSet(): void {                                                                                                                                        
     $config = $this->blockConfig;                                                                                                                                                              
     $config['display_more_link'] = TRUE;                                                                                                                                                       
     $config['more_link'] = 'https://registrar.uiowa.edu/academic-calendar';                                                                                                                    
     $config['more_text'] = 'View more';                                                                                                                                                        
                                                                                                                                                                                                
-    $sut = new AcademicDatesBlock($config, 'uiowa_maui_academic_dates', $this->plugin, $this->maui, $this->formBuilder);                                                                       
+    $sut = $this->makeBlock($config);                                                                                                                                                          
     $build = $sut->build();
+                                                                                                                                                                                               
     $this->assertArrayHasKey('more_link', $build);                                                                                                                                             
+    $this->assertSame('View more', (string) $build['more_link']['#title']);                                                                                                                    
+    $this->assertSame('https://registrar.uiowa.edu/academic-calendar', $build['more_link']['#url']->getUri());                                                                                 
   }                                                                                                                                                                                            
                                                                                                                                                                                                
   /**                                                                                                                                                                                          
-   * The more link should not render if display more link is not checked.                                                                                                                      
+   * Tests the more link is absent when disabled.                                                                                                                                              
+   *                                                                                                                                                                                           
+   * @covers ::build                                                                                                                                                                           
    */                                                                                                                                                                                          
-  public function testMoreLinkDoesRenderIfNotSet() {                                                                                                                                           
+  public function testMoreLinkDoesNotRenderIfNotSet(): void {                                                                                                                                  
     $config = $this->blockConfig;                                                                                                                                                              
     $config['display_more_link'] = FALSE;                                                                                                                                                      
-    $config['more_link'] = 'https://registrar.uiowa.edu/academic-calendar';                                                                                                                    
-    $config['more_text'] = 'View more';                                                                                                                                                        
                                                                                                                                                                                                
-    $sut = new AcademicDatesBlock($config, 'uiowa_maui_academic_dates', $this->plugin, $this->maui, $this->formBuilder);                                                                       
+    $sut = $this->makeBlock($config);                                                                                                                                                          
     $build = $sut->build();                                                                                                                                                                    
+                                                                                                                                                                                               
     $this->assertArrayNotHasKey('more_link', $build);                                                                                                                                          
   }                                                                                                                                                                                            
                                                                                                                                                                                                
   /**                                                                                                                                                                                          
-   * Data provider.                                                                                                                                                                            
+   * Provides headline placeholders and their expected replacements.                                                                                                                           
+   *                                                                                                                                                                                           
+   * @return array                                                                                                                                                                             
+   *   Each case is [input headline, expected rendered headline].                                                                                                                              
    */                                                                                                                                                                                          
-  public function placeholderProvider() {                                                                                                                                                      
+  public static function placeholderProvider(): array {                                                                                                                                        
     return [                                                                                                                                                                                   
-      ['@session'],                                                                                                                                                                            
-      ['@session Deadlines'],                                                                                                                                                                  
-      ['Upcoming Deadlines, @session'],                                                                                                                                                        
-      ['@foo @bar @session'],                                                                                                                                                                  
+      ['@session', 'Winter 2020'],                                                                                                                                                             
+      ['@session Deadlines', 'Winter 2020 Deadlines'],                                                                                                                                         
+      ['Upcoming Deadlines, @session', 'Upcoming Deadlines, Winter 2020'],                                                                                                                     
+      ['@foo @bar @session', '@foo @bar Winter 2020'],                                                                                                                                         
     ];                                                                                                                                                                                         
   }
```

</details>

**Review by pyrello — COMMENTED** [review]
I have reviewed the code and I understand all the changes. To the best of my understanding this new Unit test is testing all the same things as the Kernel test it is replacing.

**Review by cory-skeers — APPROVED** [review]
Thanks for including the line-by-line diff! Really helpful in reviewing.

I've had to do some brushing up on tests, but I also think this looks good and is a better fit for what it's actually testing, which only needs the mocks and not full loading.

**Comment by cory-skeers** [comment]
As an aside, here is the example documentation I had Claude write to my Obsidian vault when having it give this a passover as well:

<details>
<summary>
Markdown
</summary>

# Unit Testing Best Practices

This guide provides UIowa-specific guidance on writing effective PHPUnit unit tests for Drupal modules.

## What to Test (and What Not to Test)

### Test Business Logic, Not Framework Code

Following Drupal's guidance to "avoid testing Drupal internals; test your module's behaviour," focus on testing your module's unique functionality, not framework patterns.

✅ **DO test:**
- Business logic and algorithms
- Conditional logic and branching
- Data transformations
- Validation rules
- Edge cases and error handling
- Configuration handling unique to your module
- Complex calculations or string manipulations
- API response processing

❌ **DON'T test:**
- Standard Drupal patterns (`create()` methods, standard hooks)
- Framework boilerplate code
- Simple getters/setters with no logic
- Implementation details (private method internals via reflection)
- Code that's just passing through to Drupal APIs

### The create() Method Question

**General guidance: Skip testing `create()` methods** for standard dependency injection.

**Why skip it:**
- It's a standardized `ContainerInjectionInterface` pattern
- Testing it verifies Drupal's DI system, not your logic
- Unit tests should bypass it anyway by calling the constructor directly
- Adds maintenance burden when dependencies change
- Other tests implicitly verify it works

**When you MIGHT test it:**
- The `create()` method does non-standard logic beyond simple DI
- Complex factory patterns with conditional instantiation
- Service selection based on configuration or environment

**Example of what NOT to test:**
```php
// This is standard boilerplate - don't test it
public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
  return new static(
    $configuration,
    $plugin_id,
    $plugin_definition,
    $container->get('my_module.service')
  );
}
```

**References:**
- [[Drupal.org: Avoid testing Drupal internals](https://www.drupal.org/docs/develop/automated-testing/phpunit-in-drupal)](https://www.drupal.org/docs/develop/automated-testing/phpunit-in-drupal)
- [[DI facilitates bypassing create() in tests](https://medium.com/charlottes-digital-web/making-drupal-8-code-testable-limiting-the-use-of-static-methods-using-dependency-injection-faade6f3f654)](https://medium.com/charlottes-digital-web/making-drupal-8-code-testable-limiting-the-use-of-static-methods-using-dependency-injection-faade6f3f654)

## Testing Patterns for Blocks/Plugins

### Pattern: Constructor Helper Method

Instead of testing the `create()` method, use a helper that directly calls the constructor:

```php
/**
 * Builds the block under test with mocked dependencies.
 *
 * @param array $config
 *   The block configuration.
 *
 * @return \Drupal\my_module\Plugin\Block\MyBlock
 *   The configured block.
 */
protected function makeBlock(array $config): MyBlock {
  $block = new MyBlock(
    $config,
    'my_module_my_block',
    $this->pluginDefinition,
    $this->mockService1,
    $this->mockService2
  );
  $block->setStringTranslation($this->getStringTranslationStub());
  return $block;
}
```

**Benefits of this pattern:**
- Bypasses the `create()` factory method (faster, simpler)
- Makes dependencies explicit and visible
- Reduces container overhead in tests
- Follows Drupal unit testing best practices
- Makes it easy to test with different configurations

### Pattern: setUp() for Shared Mocks

Use `setUp()` to create mocks and default configurations used across multiple tests:

```php
protected function setUp(): void {
  parent::setUp();

  // Plugin definition (same for all tests)
  $this->pluginDefinition = [
    'admin_label' => 'My Block',
    'provider' => 'my_module',
    'category' => 'Custom',
  ];

  // Default configuration
  $this->defaultConfig = [
    'label' => 'Test Block',
    'option1' => TRUE,
    'option2' => 'default',
  ];

  // Mock dependencies
  $this->mockService = $this->createMock(MyServiceInterface::class);
  $this->mockCache = $this->createMock(CacheBackendInterface::class);
}
```

### Pattern: Data Providers for Multiple Scenarios

Use data providers to test multiple input/output combinations efficiently:

```php
/**
 * Tests placeholder replacement handles various formats.
 *
 * @covers ::build
 * @dataProvider placeholderProvider
 */
public function testPlaceholderReplacement(string $input, string $expected): void {
  $config = $this->defaultConfig;
  $config['headline'] = $input;

  $block = $this->makeBlock($config);
  $build = $block->build();

  $this->assertSame($expected, $build['heading']['#headline']);
}

/**
 * Provides placeholder test cases.
 *
 * @return array
 *   Test cases with [input, expected output].
 */
public static function placeholderProvider(): array {
  return [
    'simple replacement' => ['@session', 'Winter 2020'],
    'with suffix' => ['@session Deadlines', 'Winter 2020 Deadlines'],
    'with prefix' => ['Upcoming @session', 'Upcoming Winter 2020'],
    'unknown placeholders ignored' => ['@foo @session', '@foo Winter 2020'],
    'multiple same placeholder' => ['@session @session', 'Winter 2020 Winter 2020'],
  ];
}
```

**Data Provider Tips:**
- Use associative array keys to document what each test case verifies
- Group related scenarios together
- Make expected values obvious (avoid complex calculations in the provider)
- Keep providers focused (one aspect of functionality per provider)
- Use `static` for the provider method (required by PHPUnit)

## Common Unit Testing Pitfalls

### ❌ Testing Implementation Details with Reflection

```php
// BAD: Tests private property assignment (implementation detail)
public function testCreateSetsServices(): void {
  $block = MyBlock::create($container, [], 'plugin_id', $definition);

  $property = new \ReflectionProperty($block, 'myService');
  $this->assertSame($this->mockService, $property->getValue($block));
}
```

```php
// GOOD: Test behavior through public API
public function testBlockUsesServiceData(): void {
  $this->mockService->method('getData')->willReturn(['item1', 'item2']);

  $block = $this->makeBlock($this->defaultConfig);
  $build = $block->build();

  $this->assertCount(2, $build['items']);
}
```

**Why reflection is problematic:**
- Tests implementation, not behavior
- Breaks encapsulation
- Makes refactoring harder (tests break when internal structure changes)
- Doesn't verify the code actually works correctly

### ❌ Incomplete Edge Case Coverage

Always consider these scenarios:

```php
// Empty collections
$this->mockApi->method('getItems')->willReturn([]);

// NULL values
$config['optional_field'] = NULL;

// Invalid input
$config['items_to_display'] = -1;
$config['items_to_display'] = 0;
$config['items_to_display'] = 9999999;

// Missing data from external APIs
$this->mockApi->method('getSession')->willReturn(NULL);

// Malformed data
$this->mockApi->method('getItems')->willReturn([
  (object) ['id' => 1], // Missing expected 'name' property
]);
```

### ❌ Mock Methods Returning Wrong Types

```php
// BAD: Mock returns array when real service returns object
$this->mockService->method('getUser')->willReturn(['name' => 'John']);

// GOOD: Match the actual return type
$user = new \stdClass();
$user->name = 'John';
$this->mockService->method('getUser')->willReturn($user);
```

### ❌ Not Testing Cache Metadata

For blocks and entities, always verify cache metadata:

```php
public function testCacheMetadata(): void {
  $this->mockService->method('getData')->willReturn($data);

  $block = $this->makeBlock($this->defaultConfig);
  $build = $block->build();

  // Verify cache tags
  $this->assertContains('my_module:data', $build['#cache']['tags']);

  // Verify cache contexts
  $this->assertContains('url.query_args', $build['#cache']['contexts']);

  // Verify max-age
  $this->assertEquals(3600, $build['#cache']['max-age']);
}
```

## Block Plugin Unit Test Checklist

When writing unit tests for a custom block plugin, ensure coverage of:

### Configuration
- [ ] `blockValidate()` - All validation rules
- [ ] `blockSubmit()` - Configuration transformation and sanitization
- [ ] Default configuration in `defaultConfiguration()`
- [ ] Required vs optional configuration fields

### Build Logic
- [ ] `build()` - Main render array structure
- [ ] Conditional rendering (hide/show elements based on config)
- [ ] Placeholder/token replacement
- [ ] Data transformations before rendering
- [ ] Proper render array structure (`#type`, `#theme`, etc.)

### Edge Cases
- [ ] Empty state handling (no data from API/service)
- [ ] NULL or missing configuration values
- [ ] Invalid IDs (session doesn't exist, entity not found)
- [ ] Boundary values (0, negative, very large numbers)
- [ ] Missing required services or API failures

### Render Elements
- [ ] Link rendering (URLs, titles, attributes)
- [ ] Heading rendering (levels, visibility, styling)
- [ ] List items (correct count, structure, content)
- [ ] Conditional elements (more links, filters, etc.)

### Cache & Performance
- [ ] Cache tags (entity types, custom tags)
- [ ] Cache contexts (user, URL, etc.)
- [ ] Max-age settings
- [ ] Cache bubbling for nested elements

### Skip (unless non-standard)
- [ ] ~~`create()` method~~ (standard DI pattern)
- [ ] ~~Simple getters with no logic~~
- [ ] ~~Property assignments~~

## Test Organization

### File Structure
```
docroot/modules/custom/my_module/
└── tests/
    └── src/
        ├── Unit/
        │   ├── ServiceNameTest.php
        │   └── Plugin/
        │       └── Block/
        │           └── MyBlockTest.php
        ├── Kernel/
        │   └── MyModuleKernelTest.php
        └── Functional/
            └── MyModuleFunctionalTest.php
```

### Naming Conventions
- Test class: `{ClassName}Test.php`
- Test method: `test{What}{Condition}()`
- Examples:
  - `testBuildReturnsEmptyArrayWhenNoData()`
  - `testValidateRejectsInvalidUrl()`
  - `testPlaceholderReplacementWithMultipleTokens()`

### Test Method Structure

Follow the Arrange-Act-Assert pattern:

```php
public function testItemsLimitedByConfiguration(): void {
  // Arrange: Set up test data and configuration
  $items = [/* 20 items */];
  $this->mockApi->method('getItems')->willReturn($items);
  $config = $this->defaultConfig;
  $config['items_to_display'] = 5;

  // Act: Execute the code under test
  $block = $this->makeBlock($config);
  $build = $block->build();

  // Assert: Verify the outcome
  $this->assertCount(5, $build['items']);
}
```

## Running Tests

See [[[Testing](https://github.com/uiowa/uiowa/pull/Testing)]] for commands to run tests.

### Run specific test class
```bash
ddev phpunit docroot/modules/custom/my_module/tests/src/Unit/MyClassTest.php
```

### Run with coverage (slow)
```bash
ddev phpunit --coverage-html /tmp/coverage docroot/modules/custom/my_module/
```

### Run specific test method
```bash
ddev phpunit --filter testMethodName docroot/modules/custom/my_module/tests/
```

## Additional Resources

- [[[Testing](https://github.com/uiowa/uiowa/pull/Testing)]] - Main testing documentation
- [[[Development Workflow](https://github.com/uiowa/uiowa/pull/Development%20Workflow)]] - Daily development tasks including testing
- [[Drupal.org: PHPUnit in Drupal](https://www.drupal.org/docs/develop/automated-testing/phpunit-in-drupal)](https://www.drupal.org/docs/develop/automated-testing/phpunit-in-drupal)
- [[Drupal.org: Types of Tests](https://www.drupal.org/docs/develop/automated-testing/types-of-tests)](https://www.drupal.org/docs/develop/automated-testing/types-of-tests)
- [[PHPUnit Documentation](https://phpunit.de/documentation.html)](https://phpunit.de/documentation.html)

---

**Last updated:** 2026-05-29

</details>

**Comment by pyrello** [comment]
> As an aside, here is the example documentation I had Claude write

Nobody is going to accuse Claude of not providing enough details
